File: src/tzx/wire.ts
// Decoder for the byte format the Rust core answers in; `core/src/wire.rs` is
// the encoder and documents the layout. Fields are read in the order the block
// literals below declare them, which is the order `parser.ts` used to build
// them in, so key order (and therefore anything comparing JSON) is unchanged.
import { Reader, Writer } from './bytes';
import { ArchiveEntry, Block, HardwareEntry, isUnknown, newUid, ParsedTape, PilotRun, SelectEntry, SymDef } from './types';

export const WIRE_VERSION = 1;
const UNKNOWN_TAG = 0xff;

/** The header every answer starts with; throws on an error payload. */
function readHeader(r: Reader): void {
  const version = r.u8();
  if (version !== WIRE_VERSION) {
    throw new Error(`Tape core speaks wire format ${version}, this build expects ${WIRE_VERSION}`);
  }
  if (r.u8() === 1) throw new Error(str(r));
}

/** A byte string answer: a serialized tape or block. */
export function decodeBytes(buf: Uint8Array): Uint8Array {
  const r = new Reader(buf);
  readHeader(r);
  return bytes(r);
}

/** A TAP answer: the bytes plus the indices of the blocks left out. */
export function decodeTap(buf: Uint8Array): { bytes: Uint8Array; skipped: number[] } {
  const r = new Reader(buf);
  readHeader(r);
  const out = bytes(r);
  const skipped: number[] = [];
  for (let n = r.u32(); n > 0; n--) skipped.push(r.u32());
  return { bytes: out, skipped };
}

/** A TZX version answer. */
export function decodeVersion(buf: Uint8Array): { major: number; minor: number } {
  const r = new Reader(buf);
  readHeader(r);
  return { major: r.u8(), minor: r.u8() };
}

export function decodeTape(buf: Uint8Array): ParsedTape {
  const r = new Reader(buf);
  readHeader(r);
  const major = r.u8();
  const minor = r.u8();
  const warnings: string[] = [];
  for (let n = r.u32(); n > 0; n--) warnings.push(str(r));
  const blocks: Block[] = [];
  for (let n = r.u32(); n > 0; n--) blocks.push(block(r));
  return { blocks, major, minor, warnings };
}

function str(r: Reader): string {
  return r.str(r.u32());
}

function bytes(r: Reader): Uint8Array {
  return r.bytes(r.u32());
}

function u16s(r: Reader): number[] {
  const out: number[] = [];
  for (let n = r.u32(); n > 0; n--) out.push(r.u16());
  return out;
}

function symDefs(r: Reader): SymDef[] {
  const out: SymDef[] = [];
  for (let n = r.u32(); n > 0; n--) out.push({ flags: r.u8(), pulses: u16s(r) });
  return out;
}

function block(r: Reader): Block {
  const uid = newUid();
  const tag = r.u8();
  if (tag === UNKNOWN_TAG) {
    const id = r.u8();
    return { uid, id, unknown: true, raw: bytes(r) };
  }
  const id = tag;
  switch (id) {
    case 0x10:
      return { uid, id, pause: r.u16(), data: bytes(r) };
    case 0x11:
      return {
        uid, id, pilot: r.u16(), sync1: r.u16(), sync2: r.u16(), zero: r.u16(), one: r.u16(),
        pilotLen: r.u16(), usedBits: r.u8(), pause: r.u16(), data: bytes(r),
      };
    case 0x12:
      return { uid, id, pulseLen: r.u16(), count: r.u16() };
    case 0x13:
      return { uid, id, pulses: u16s(r) };
    case 0x14:
      return { uid, id, zero: r.u16(), one: r.u16(), usedBits: r.u8(), pause: r.u16(), data: bytes(r) };
    case 0x15:
      return { uid, id, tstates: r.u16(), pause: r.u16(), usedBits: r.u8(), data: bytes(r) };
    case 0x18:
      return {
        uid, id, pause: r.u16(), sampleRate: r.u32(), compression: r.u8(), pulseCount: r.u32(),
        data: bytes(r),
      };
    case 0x19: {
      const pause = r.u16();
      const totp = r.u32();
      const npp = r.u8();
      const pilotSymbols = symDefs(r);
      const pilotStream: PilotRun[] = [];
      for (let n = r.u32(); n > 0; n--) pilotStream.push({ symbol: r.u8(), reps: r.u16() });
      const totd = r.u32();
      const npd = r.u8();
      const dataSymbols = symDefs(r);
      return { uid, id, pause, totp, npp, pilotSymbols, pilotStream, totd, npd, dataSymbols, data: bytes(r) };
    }
    case 0x20:
      return { uid, id, pause: r.u16() };
    case 0x21:
      return { uid, id, name: str(r) };
    case 0x22:
      return { uid, id };
    case 0x23:
      return { uid, id, offset: r.i16() };
    case 0x24:
      return { uid, id, count: r.u16() };
    case 0x25:
      return { uid, id };
    case 0x26: {
      const offsets: number[] = [];
      for (let n = r.u32(); n > 0; n--) offsets.push(r.i16());
      return { uid, id, offsets };
    }
    case 0x27:
      return { uid, id };
    case 0x28: {
      const entries: SelectEntry[] = [];
      for (let n = r.u32(); n > 0; n--) entries.push({ offset: r.i16(), text: str(r) });
      return { uid, id, entries };
    }
    case 0x2a:
      return { uid, id };
    case 0x2b:
      return { uid, id, level: r.u8() };
    case 0x30:
      return { uid, id, text: str(r) };
    case 0x31:
      return { uid, id, time: r.u8(), text: str(r) };
    case 0x32: {
      const entries: ArchiveEntry[] = [];
      for (let n = r.u32(); n > 0; n--) entries.push({ type: r.u8(), text: str(r) });
      return { uid, id, entries };
    }
    case 0x33: {
      const entries: HardwareEntry[] = [];
      for (let n = r.u32(); n > 0; n--) entries.push({ type: r.u8(), id: r.u8(), info: r.u8() });
      return { uid, id, entries };
    }
    case 0x35:
      return { uid, id, ident: str(r), data: bytes(r) };
    case 0x5a:
      return { uid, id, raw: bytes(r) };
    default:
      throw new Error(`Tape core sent block tag ${id.toString(16)}, which this build does not know`);
  }
}

// ---- encoding, for the calls that send blocks to the core ------------------

/**
 * A block list in the format `core/src/wire.rs` decodes.
 *
 * With `withData: false` the byte payloads are left out — the whole point of a
 * tape is its data, so sending it across for a call that only looks at block
 * types costs more than the call. Only for entry points that provably ignore
 * the data (`core_required_version`, `core_save_version`); the differential
 * tests compare them against the reference on blocks that do carry data, so a
 * core that started reading it would fail there.
 */
export function encodeBlocks(blocks: Block[], opts: { withData?: boolean } = {}): Uint8Array {
  const w = new Writer();
  const put = opts.withData === false ? skipBytes : putBytes;
  w.u32(blocks.length);
  for (const b of blocks) writeBlock(w, b, put);
  return w.toUint8Array();
}

/** How a block's byte payloads go on the wire. */
type PutBytes = (w: Writer, b: Uint8Array) => void;

function putBytes(w: Writer, b: Uint8Array): void {
  w.u32(b.length);
  w.bytes(b);
}

function skipBytes(w: Writer, _b: Uint8Array): void {
  w.u32(0);
}

function putStr(w: Writer, s: string): void {
  w.u32(s.length);
  w.str(s);
}

function putU16s(w: Writer, v: number[]): void {
  w.u32(v.length);
  for (const n of v) w.u16(n);
}

function putSymDefs(w: Writer, defs: SymDef[]): void {
  w.u32(defs.length);
  for (const d of defs) {
    w.u8(d.flags);
    putU16s(w, d.pulses);
  }
}

function writeBlock(w: Writer, b: Block, putBytes: PutBytes): void {
  if (isUnknown(b)) {
    w.u8(UNKNOWN_TAG);
    w.u8(b.id);
    putBytes(w, b.raw);
    return;
  }
  w.u8(b.id);
  switch (b.id) {
    case 0x10:
      w.u16(b.pause);
      putBytes(w, b.data);
      break;
    case 0x11:
      for (const v of [b.pilot, b.sync1, b.sync2, b.zero, b.one, b.pilotLen]) w.u16(v);
      w.u8(b.usedBits);
      w.u16(b.pause);
      putBytes(w, b.data);
      break;
    case 0x12:
      w.u16(b.pulseLen);
      w.u16(b.count);
      break;
    case 0x13:
      putU16s(w, b.pulses);
      break;
    case 0x14:
      w.u16(b.zero);
      w.u16(b.one);
      w.u8(b.usedBits);
      w.u16(b.pause);
      putBytes(w, b.data);
      break;
    case 0x15:
      w.u16(b.tstates);
      w.u16(b.pause);
      w.u8(b.usedBits);
      putBytes(w, b.data);
      break;
    case 0x18:
      w.u16(b.pause);
      w.u32(b.sampleRate);
      w.u8(b.compression);
      w.u32(b.pulseCount);
      putBytes(w, b.data);
      break;
    case 0x19:
      w.u16(b.pause);
      w.u32(b.totp);
      w.u8(b.npp);
      putSymDefs(w, b.pilotSymbols);
      w.u32(b.pilotStream.length);
      for (const run of b.pilotStream) {
        w.u8(run.symbol);
        w.u16(run.reps);
      }
      w.u32(b.totd);
      w.u8(b.npd);
      putSymDefs(w, b.dataSymbols);
      putBytes(w, b.data);
      break;
    case 0x20:
      w.u16(b.pause);
      break;
    case 0x21:
      putStr(w, b.name);
      break;
    case 0x22:
      break;
    case 0x23:
      w.i16(b.offset);
      break;
    case 0x24:
      w.u16(b.count);
      break;
    case 0x25:
      break;
    case 0x26:
      w.u32(b.offsets.length);
      for (const o of b.offsets) w.i16(o);
      break;
    case 0x27:
      break;
    case 0x28:
      w.u32(b.entries.length);
      for (const e of b.entries) {
        w.i16(e.offset);
        putStr(w, e.text);
      }
      break;
    case 0x2a:
      break;
    case 0x2b:
      w.u8(b.level);
      break;
    case 0x30:
      putStr(w, b.text);
      break;
    case 0x31:
      w.u8(b.time);
      putStr(w, b.text);
      break;
    case 0x32:
      w.u32(b.entries.length);
      for (const e of b.entries) {
        w.u8(e.type);
        putStr(w, e.text);
      }
      break;
    case 0x33:
      w.u32(b.entries.length);
      for (const e of b.entries) {
        w.u8(e.type);
        w.u8(e.id);
        w.u8(e.info);
      }
      break;
    case 0x35:
      putStr(w, b.ident);
      putBytes(w, b.data);
      break;
    case 0x5a:
      putBytes(w, b.raw);
      break;
  }
}

File: core/src/wire.rs
//! The binary format the wasm build hands parsed tapes to JavaScript in.
//!
//! `src/tzx/wire.ts` decodes it. Both ends are hand written, so the format is
//! deliberately dull: little-endian, fixed widths that match the block model,
//! no alignment and no compression. Fields appear in the order the TypeScript
//! object literals in `parser.ts` declare them, so the decoded objects come out
//! with the same key order as before.
//!
//! ```text
//! u8  WIRE_VERSION
//! u8  status        0 ok, 1 error
//! error:  str message
//! ok:     u8 major, u8 minor
//!         u32 count, then that many str   (warnings)
//!         u32 count, then that many block:
//!           u8 tag        the block ID, or UNKNOWN_TAG
//!           UNKNOWN_TAG:  u8 id, bytes raw
//!           otherwise:    the fields of that block type, in declaration order
//! ```
//!
//! `str` and `bytes` are both a u32 length and that many bytes; `str` holds the
//! Latin-1 bytes of the TZX text, which the decoder turns back into a JS string
//! one char per byte.

use crate::parser::ParsedTape;
use crate::types::{Block, Body, SymDef};

pub const WIRE_VERSION: u8 = 1;
const UNKNOWN_TAG: u8 = 0xff;

pub fn encode_tape(t: &ParsedTape) -> Vec<u8> {
    let mut w = Vec::with_capacity(4096);
    w.push(WIRE_VERSION);
    w.push(0); // ok
    w.push(t.major);
    w.push(t.minor);
    u32v(&mut w, t.warnings.len());
    for warning in &t.warnings {
        string(&mut w, warning);
    }
    u32v(&mut w, t.blocks.len());
    for b in &t.blocks {
        block(&mut w, b);
    }
    w
}

pub fn encode_error(message: &str) -> Vec<u8> {
    let mut w = Vec::with_capacity(message.len() + 8);
    w.push(WIRE_VERSION);
    w.push(1); // error
    string(&mut w, message);
    w
}

fn block(w: &mut Vec<u8>, b: &Block) {
    if let Body::Unknown { id, raw } = &b.body {
        w.push(UNKNOWN_TAG);
        w.push(*id);
        bytes(w, raw);
        return;
    }
    w.push(b.id());
    match &b.body {
        Body::Standard { pause, data } => {
            u16v(w, *pause);
            bytes(w, data);
        }
        Body::Turbo { pilot, sync1, sync2, zero, one, pilot_len, used_bits, pause, data } => {
            for v in [*pilot, *sync1, *sync2, *zero, *one, *pilot_len] {
                u16v(w, v);
            }
            w.push(*used_bits);
            u16v(w, *pause);
            bytes(w, data);
        }
        Body::PureTone { pulse_len, count } => {
            u16v(w, *pulse_len);
            u16v(w, *count);
        }
        Body::PulseSeq { pulses } => u16s(w, pulses),
        Body::PureData { zero, one, used_bits, pause, data } => {
            u16v(w, *zero);
            u16v(w, *one);
            w.push(*used_bits);
            u16v(w, *pause);
            bytes(w, data);
        }
        Body::Direct { tstates, pause, used_bits, data } => {
            u16v(w, *tstates);
            u16v(w, *pause);
            w.push(*used_bits);
            bytes(w, data);
        }
        Body::Csw { pause, sample_rate, compression, pulse_count, data } => {
            u16v(w, *pause);
            u32v(w, *sample_rate as usize);
            w.push(*compression);
            u32v(w, *pulse_count as usize);
            bytes(w, data);
        }
        Body::Generalized {
            pause,
            totp,
            npp,
            pilot_symbols,
            pilot_stream,
            totd,
            npd,
            data_symbols,
            data,
        } => {
            u16v(w, *pause);
            u32v(w, *totp as usize);
            w.push(*npp);
            sym_defs(w, pilot_symbols);
            u32v(w, pilot_stream.len());
            for run in pilot_stream {
                w.push(run.symbol);
                u16v(w, run.reps);
            }
            u32v(w, *totd as usize);
            w.push(*npd);
            sym_defs(w, data_symbols);
            bytes(w, data);
        }
        Body::Pause { pause } => u16v(w, *pause),
        Body::GroupStart { name } => string(w, name),
        Body::GroupEnd => {}
        Body::Jump { offset } => u16v(w, *offset as u16),
        Body::LoopStart { count } => u16v(w, *count),
        Body::LoopEnd => {}
        Body::Call { offsets } => {
            u32v(w, offsets.len());
            for o in offsets {
                u16v(w, *o as u16);
            }
        }
        Body::Return => {}
        Body::Select { entries } => {
            u32v(w, entries.len());
            for e in entries {
                u16v(w, e.offset as u16);
                string(w, &e.text);
            }
        }
        Body::Stop48 => {}
        Body::SignalLevel { level } => w.push(*level),
        Body::Text { text } => string(w, text),
        Body::Message { time, text } => {
            w.push(*time);
            string(w, text);
        }
        Body::Archive { entries } => {
            u32v(w, entries.len());
            for e in entries {
                w.push(e.kind);
                string(w, &e.text);
            }
        }
        Body::Hardware { entries } => {
            u32v(w, entries.len());
            for e in entries {
                w.push(e.kind);
                w.push(e.id);
                w.push(e.info);
            }
        }
        Body::Custom { ident, data } => {
            string(w, ident);
            bytes(w, data);
        }
        Body::Glue { raw } => bytes(w, raw),
        Body::Unknown { .. } => unreachable!("handled above"),
    }
}

fn sym_defs(w: &mut Vec<u8>, defs: &[SymDef]) {
    u32v(w, defs.len());
    for d in defs {
        w.push(d.flags);
        u16s(w, &d.pulses);
    }
}

fn u16v(w: &mut Vec<u8>, v: u16) {
    w.extend_from_slice(&v.to_le_bytes());
}

fn u32v(w: &mut Vec<u8>, v: usize) {
    w.extend_from_slice(&(v as u32).to_le_bytes());
}

fn u16s(w: &mut Vec<u8>, v: &[u16]) {
    u32v(w, v.len());
    for n in v {
        u16v(w, *n);
    }
}

fn bytes(w: &mut Vec<u8>, b: &[u8]) {
    u32v(w, b.len());
    w.extend_from_slice(b);
}

fn string(w: &mut Vec<u8>, s: &str) {
    let b = crate::bytes::string_to_latin1(s);
    bytes(w, &b);
}
